### DISCUSSION TITLE
Workouts Monitoring - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#6739] [Docs]
# Workouts Monitoring - Ultralytics YOLOv8 Docs

Workouts Monitoring Using Ultralytics YOLOv8

https://docs.ultralytics.com/guides/workouts-monitoring/

**Comment by giter915** [comment]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[1], line 23
     21     frame_count += 1
     22     results = model.predict(im0, verbose=False)
---> 23     im0 = gym_object.start_counting(im0, results, frame_count)
     25 cv2.destroyAllWindows()

File ~/anaconda3/envs/y/lib/python3.9/site-packages/ultralytics/solutions/ai_gym.py:81, in AIGym.start_counting(self, im0, results, frame_count)
     79 for ind, k in enumerate(reversed(self.keypoints)):
     80     if self.pose_type == 'pushup' or self.pose_type == 'pullup':
---> 81         self.angle[ind] = self.annotator.estimate_pose_angle(k[int(self.kpts_to_check[0])].cpu(),
     82                                                              k[int(self.kpts_to_check[1])].cpu(),
     83                                                              k[int(self.kpts_to_check[2])].cpu())
     84         self.im0 = self.annotator.draw_specific_points(k, self.kpts_to_check, shape=(640, 640), radius=10)
     86     if self.pose_type == 'abworkout':

IndexError: list assignment index out of range

**Reply by RizwanMunawar** [reply]
Hey @giter915, could you please double-check the video dimensions? Currently, the video aspect ratio of 16:9 is supported! We are actively addressing generic resolution issues and will provide an update soon!

Thanks
Ultralytics Team!

**Reply by gvzdv** [reply]
It crashes when it even momentarily loses one or more keypoints. You can override it by implementing this change in `ai_gym.py`:

    `def start_counting(self, im0, results, frame_count, fps):
        """
        Function used to count the gym steps
        Args:
            im0 (ndarray): Current frame from the video stream.
            results: Pose estimation data
            frame_count: store current frame count
        """
        self.im0 = im0
        if frame_count == 1:
            self.count = [0] * len(results[0])
            self.angle = [0] * len(results[0])
            self.stage = ["-" for _ in results[0]]
        self.keypoints = results[0].keypoints.data
        self.annotator = Annotator(im0, line_width=2)

        num_keypoints = len(results[0])

        # Resize self.angle, self.count, and self.stage if the number of keypoints has changed
        if len(self.angle) != num_keypoints:
            self.angle = [0] * num_keypoints
            self.count = [0] * num_keypoints
            self.stage = ["-" for _ in range(num_keypoints)]`

**Reply by RizwanMunawar** [reply]
Thank you, @gvzdv, for bringing attention to this issue. Could you kindly submit a pull request (PR) that can resolve this issue? We encourage community contributions and want everyone to receive due credit for their efforts. 

We eagerly await your PR.

Best regards,
Ultralytics Team

**Reply by gvzdv** [reply]
Hi @RizwanMunawar,

No worries, happy to help improve the awesome product!

I've just submitted the PR. Feel free to let me know if it needs any changes.

Kind regards,
Mike Gvozdev

**Reply by RizwanMunawar** [reply]
@gvzdv Thanks for the PR, I just left few comments. Kindly review and we can merge the PR :)

Thanks
Ultralytics Team!

**Reply by saadndm** [reply]
Why was the solution implemented in this PR removed in this https://github.com/ultralytics/ultralytics/commit/dcb953bc4139e560a6878d255afdb70a5891ec73 commit? I am getting the same error now because the fix was removed

**Reply by pderrenger** [reply]
@BridgeSenseDev hey there,

Thanks for bringing this to our attention! 🚀 The change in the commit you pointed out was part of a larger refactor aiming to improve the system's overall stability and performance. However, your feedback is incredibly valuable, and we might have overlooked how this impacts users dealing with the same issue the PR originally addressed.

Could you please provide details about the error you're encountering? If possible, a snippet of the code or steps to reproduce the issue would be beneficial 📝. We aim to ensure our solutions work seamlessly for a wide range of applications, including workouts monitoring like the one mentioned in our docs. 

We're here to help and ensure you have a smooth experience using Ultralytics YOLOv8 for your awesome projects!

**Comment by Anto5040** [comment]
I tried it with the webcam by changing ```cap = cv2.VideoCapture("path/to/video/file.mp4")``` to ```cap = cv2.VideoCapture(0)
cap.set(3, 640)
cap.set(4, 480)```
But got this error:
Traceback (most recent call last):
  File "path/workout_example.py", line 25, in <module>
    im0 = gym_object.start_counting(im0, results, frame_count)
  File "conda_env/lib/python3.9/site-packages/ultralytics/solutions/ai_gym.py", line 81, in start_counting
    self.angle[ind] = self.annotator.estimate_pose_angle(k[int(self.kpts_to_check[0])].cpu(),
IndexError: list assignment index out of range

**Reply by RizwanMunawar** [reply]
This issue has been resolved now in latest Ultralytics released!!

Thanks

**Comment by Anto5040** [comment]
Some really simple suggestions: Other movements could be easily implemented. Squats, for example, could be counted easily with the knee and hip angles. A first aproach could be to track if the hip joint is lower than the knee joint, but some more robust methods should be tried in the future.

**Reply by pderrenger** [reply]
@Anto5040 thank you for your suggestion! It's great to hear ideas for enhancing the capabilities of our models. You're correct that tracking specific joint angles could be a viable method for counting exercises like squats. We're always looking to improve and expand the functionality of our models, and incorporating pose and keypoint estimation for workout monitoring is indeed a promising direction.

If you're interested in implementing this yourself, you might want to explore the pose estimation capabilities of YOLOv8, which can detect keypoints of the human body. You could use these keypoints to calculate angles and track movements over time.

For more details on how to use pose estimation with YOLOv8, you can refer to our Pose/Keypoint Estimation documentation. If you have further insights or would like to contribute to the project, we encourage you to engage with the community. Your contributions could be valuable to others looking to apply YOLOv8 in similar use cases. 😊🚀
